File: Python/Heap_PriorityQueue_Medium_KthSmallestElementInASorted Matrix_378.py
class Solution:
    def kthSmallest(self, matrix: list[list[int]], k: int) -> int:
        pq = []

        for row in range(len(matrix)):
            pq.append((matrix[row][0], row, 0))

        heapq.heapify(pq)

        for step in range(k - 1):
            val, row, col = heapq.heappop(pq)

            if col + 1 < len(matrix[0]):
                heapq.heappush(pq, (matrix[row][col + 1], row, col + 1))

        return pq[0][0]

File: C++/Heap_PriorityQueue_Medium_KthSmallestElementInASorted Matrix_378.cpp
class Solution {
public:
    int kthSmallest(vector<vector<int>>& matrix, int k) 
    {
        priority_queue<vector<int>, vector<vector<int>>, greater<vector<int>>> pq;

        for(int i = 0; i < matrix.size(); i++)
        {
            pq.push({matrix[i][0], i, 0});
        }

        for(int i = 1; i < k; i++)
        {
            auto curr = pq.top();
            pq.pop();

            int r = curr[1];
            int c = curr[2];

            if(c + 1 < matrix[0].size())
            {
                pq.push({matrix[r][c + 1], r, c + 1});
            }
        }

        return pq.top()[0];
    }
};